torch/distributions/kl.py

```diff
@@ -172,14 +172,14 @@ def _kl_beta_beta(p, q):
 @register_kl(Dirichlet, Dirichlet)
 def _kl_dirichlet_dirichlet(p, q):
     # From http://bariskurt.com/kullback-leibler-divergence-between-two-dirichlet-and-beta-distributions/
-    sum_p_alpha = p.alpha.sum(-1)
-    sum_q_alpha = q.alpha.sum(-1)
+    sum_p_alpha = p.alpha.sum(0)
+    sum_q_alpha = q.alpha.sum(0)
     t1 = torch.lgamma(sum_p_alpha)
     t2 = torch.lgamma(sum_q_alpha)
-    t3 = p.alpha.lgamma().sum(-1)
-    t4 = q.alpha.lgamma().sum(-1)
+    t3 = p.alpha.lgamma().sum(0)
+    t4 = q.alpha.lgamma().sum(0)
     t5 = (p.alpha - q.alpha) * (p.alpha.digamma() - sum_p_alpha.digamma())
```

Comment:
This line looks like it has a math bug: you're subtracting `sum_p_alpha.digamma()` of shape `(4, 4)` from `p.alpha.digamma()` of shape `(4, 4, 3)`.

Reply:
Could you suggest how to circumvent this issue? I tried `expand_as`, which is not working.

Reply:
Going from [this derivation](http://bariskurt.com/kullback-leibler-divergence-between-two-dirichlet-and-beta-distributions/), I believe you can
```py
p_alpha_sum = p.alpha.sum(-1)
return (p_alpha_sum.lgamma() - p.alpha.lgamma().sum(-1) -
    q.alpha.sum(-1).lgamma() + q.alpha.lgamma().sum(-1) +
    ((p.alpha - q.alpha) * (p.alpha.digamma() - q.alpha.digamma)).sum(-1))
```

Reply:
I think this formula is incorrect:
```python
return (p_alpha_sum.lgamma() - p.alpha.lgamma().sum(-1) - q.alpha.sum(-1).lgamma() + q.alpha.lgamma().sum(-1) + ((p.alpha - q.alpha) * (p.alpha.digamma() - p_alpha_sum.digamma())).sum(-1)
```
This falls back to the broadcasting issue. 
I thought of this (which is possibly incorrect as well):
```
t1 = sum_p_alpha.lgamma() - p.alpha.lgamma().sum(-1)
t2 = sum_q_alpha.lgamma() - q.alpha.lgamma().sum(-1)
t3 = ((p.alpha - q.alpha) * p.alpha.digamma()).sum(-1)
t4 = sum_p_alpha * (p.alpha - q.alpha).sum(-1)
return t1 - t2 + t3 - t4
```
Unfortunately, the tests fail for `set_rng_seed(0)`. One of the values are negative. No broadcasting problems, however.
```
Expected (Monte Carlo): 
 0.0000  5.1686  1.0550  0.4030
 4.0572  0.0000  0.7030  1.6659
 1.3686  1.2376  0.0000  0.3319
 0.4208  2.2028  0.2097  0.0000
[torch.DoubleTensor of size 4x4]

Actual (analytic): 
 0.0000  5.3725  1.0450  0.7444
 4.0794  0.0000  0.6892  1.9651
 1.5648  1.3234  0.0000  0.6453
 0.1315  1.8968 -0.1031  0.0000
[torch.DoubleTensor of size 4x4]
```

Reply:
You're right. I think we could fix this with an `.unsqueeze(-1)` on the last line:
```py
return (p_alpha_sum.lgamma() - p.alpha.lgamma().sum(-1) -
    q.alpha.sum(-1).lgamma() + q.alpha.lgamma().sum(-1) +
    ((p.alpha - q.alpha) * (p.alpha.digamma() - p_alpha_sum.digamma().unsqueeze(-1))).sum(-1))
```